controller/konnect/ops/ops_kongkey.go

```diff
@@ -0,0 +1,139 @@
+package ops
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+
+	sdkkonnectcomp "github.com/Kong/sdk-konnect-go/models/components"
+	sdkkonnectops "github.com/Kong/sdk-konnect-go/models/operations"
+	sdkkonnecterrs "github.com/Kong/sdk-konnect-go/models/sdkerrors"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+	"github.com/kong/kubernetes-configuration/pkg/metadata"
+)
+
+// createKey creates a KongKey in Konnect.
+// It sets the KonnectID and the Programmed condition in the KongKey status.
+func createKey(
+	ctx context.Context,
+	sdk KeysSDK,
+	key *configurationv1alpha1.KongKey,
+) error {
+	cpID := key.GetControlPlaneID()
+	if cpID == "" {
+		return fmt.Errorf("can't create %T %s without a Konnect ControlPlane ID", key, client.ObjectKeyFromObject(key))
+	}
+
+	resp, err := sdk.CreateKey(ctx,
+		cpID,
+		kongKeyToKeyInput(key),
+	)
+
+	// TODO: handle already exists
+	// Can't adopt it as it will cause conflicts between the controller
+	// that created that entity and already manages it, hm
+	if errWrap := wrapErrIfKonnectOpFailed(err, CreateOp, key); errWrap != nil {
+		SetKonnectEntityProgrammedConditionFalse(key, "FailedToCreate", errWrap.Error())
+		return errWrap
+	}
+
+	key.Status.Konnect.SetKonnectID(*resp.Key.ID)
+	SetKonnectEntityProgrammedCondition(key)
+
+	return nil
+}
+
+// updateKey updates a KongKey in Konnect.
+// The KongKey must have a KonnectID set in its status.
+// It returns an error if the KongKey does not have a KonnectID.
+func updateKey(
+	ctx context.Context,
+	sdk KeysSDK,
+	key *configurationv1alpha1.KongKey,
+) error {
+	cpID := key.GetControlPlaneID()
+	if cpID == "" {
+		return fmt.Errorf("can't update %T without a ControlPlaneID", key)
+	}
+
+	_, err := sdk.UpsertKey(ctx,
+		sdkkonnectops.UpsertKeyRequest{
+			ControlPlaneID: cpID,
+			KeyID:          key.GetKonnectStatus().GetKonnectID(),
+			Key:            kongKeyToKeyInput(key),
+		},
+	)
+
+	// TODO: handle already exists
+	// Can't adopt it as it will cause conflicts between the controller
+	// that created that entity and already manages it, hm
+	if errWrap := wrapErrIfKonnectOpFailed(err, UpdateOp, key); errWrap != nil {
+		SetKonnectEntityProgrammedConditionFalse(key, "FailedToUpdate", errWrap.Error())
+		return errWrap
+	}
```

Comment:
Can we fall back to create the key if it does not exist on update? Just like we do it for other object e.g. service in https://github.com/Kong/gateway-operator/blob/b449bc72d51a14c26bab4fe9bbbffefb13b93c2c/controller/konnect/ops/ops_kongservice.go#L75-L103